getpaid/models.py

```diff
@@ -23,133 +18,290 @@
+    ('accepted', _('accepted')),
+    ('rejected', _('rejected')),
+    ('check', _('check')),
+)
 
-class PaymentManager(models.Manager):
-    def get_queryset(self):
-        return super(PaymentManager, self).get_queryset().select_related('order')
 
+class AbstractOrder(models.Model):
+    class Meta:
+        abstract = True
 
-@python_2_unicode_compatible
-class PaymentFactory(models.Model, AbstractMixin):
-    """
-    This is an abstract class that defines a structure of Payment model that will be
-    generated dynamically with one additional field: ``order``
-    """
+    def get_redirect_url(self, *args, success=None, **kwargs):
+        """
+        Method used to determine the final url the client should see after
+        returning from gateway. Client will be redirected to this url after
+        backend handled the original callback (i.e. updated payment status)
+        and only if SUCCESS_URL or FAILURE_URL settings are NOT set.
+        By default it returns the result of `get_absolute_url`
+        """
+        return self.get_absolute_url()
+
+    def get_absolute_url(self):
+        """
+        Standard method recommended in Django docs. It should return
+        the URL to see details of particular Order.
+        """
+        raise NotImplementedError
+
+    def is_ready_for_payment(self):
+        """Most of the validation is made in PaymentMethodForm using but if you
+        need any extra validation. For example you most probably want to disable
+        making another payment for order that is already paid."""
+        return True
+
+    def get_items(self):
+        """
+        There are backends that require some sort of item list to be attached
+        to the payment. But it's up to you if the list is real or contains only
+        one item called "Payment for stuff in {myshop}" ;)
+        :return: List of {"name": str, "quantity": Decimal, "unit_price": Decimal} dicts.
+        """
+        return [{
+            'name': self.get_description(),
+            'quantity': 1,
+            'unit_price': self.get_total_amount(),
+        }]
+
+    def get_total_amount(self):
+        """
+        This method must return the total value of the Order.
+        :return: Decimal object
+        """
+        raise NotImplementedError
+
+    def get_user_info(self):
+        """
+        This method should return dict with necessary user info.
+        For most backends email should be sufficient.
+        Expected field names: `email`, `first_name`, `last_name`, `phone`
+        """
+        raise NotImplementedError
+
+    def get_description(self):
+        """
+        :return: Description of the Order. Should return the value of appropriate field.
+        """
+        raise NotImplementedError
+
+
+class AbstractPayment(models.Model):
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    order = models.ForeignKey(swapper.get_model_name('getpaid', 'Order'), verbose_name=_("order"),
```

Comment:
Maybe the relation between an order and payment should also be reviewed. Isn't that the *Order* that can be “partially paid” if the sum of the payments are less than the total amount? We have a Order-Payment pair which seems one-to-one (correct me if I'm work), but we don't have a model that would represent a single balance change (if the payment is done in batches for some reason).